File: .notes/curryings/Python/curryings-introspection.py

"""
introspection ver.
"""

fixedpoint = lambda f: (lambda x: x (x)) (lambda y: f(lambda *args: y (y) (*args))) ;

curry = lambda func: fixedpoint (lambda curried: 
	lambda *args: (lambda x: curried (* (args + (x,)))) if 
	(len (args) != func.__code__.co_argcount) else 
	func (*args)) ;

"""
# e.g. def
@curry
def prod3 (x, y, z):
	return x + y + z

print (prod3 (3) (4) (5)) # 12

# e.g. lambda
prod_3 = curry (lambda x,y,z: x + y + z)
print (prod_3 (6) (7) (8)) # 21
"""

File: .notes/curryings/R/curryings-introspection.R

#' introspection ver.

fixedpoint = \ (f) (\ (x) (x) (x)
	) (\ (y) (f) (\ (...) (y) (y) (...))) ;

curry = \ (func) fixedpoint (\ (curried) 
	\ (...) if (base::length (base::list (...)) != base::length (base::formals (func))) 
		(\ (x) curried |> base::do.call (base::list (...) |> base::c (x))) else 
		(func) (...)) ;

#' @examples
#' 
#' define `aaba = curry (\ (x, y, z) x - y - z)`
#' then run `aaba (1) (2) (3)` will get `[1] -4`
#' 
